src/passes/unloadingAssignment.ts

```diff
@@ -63,4 +82,48 @@ export class UnloadingAssignment extends ASTMapper {
       this.dispatchVisit(compoundAssignment, ast);
     }
   }
+
+  // node is the lvalue of a compound assignment
+  // this function extracts subexpressions that are not guaranteed to produce the same result every time
+  // to constants to evaluate once before the assignment
+  // this is needed because extracting to a 'constant' and then assigning to it would not modify the correct variable
+  // we also extract non-trivial expressions so they're only calculated once, though this is not technically necessary
+  // an array of created statements are returned so they can be visited
+  private extractUnstableSubexpressions(
+    node: Expression,
+    ast: AST,
+  ): VariableDeclarationStatement[] {
+    if (node instanceof IndexAccess) {
+      // a[i++] += 1 -> x = i++; a[x] = a[x] + 1
+      assert(
+        node.vIndexExpression !== undefined,
+        `Unexpected empty index access in compound assignment ${printNode(node)}`,
+      );
+      if (!(node.vIndexExpression instanceof Literal)) {
+        const decl = this.extractIndividualSubExpression(node.vIndexExpression, ast);
+        return [decl, ...this.extractUnstableSubexpressions(node.vBaseExpression, ast)];
+      }
+    } else if (node instanceof MemberAccess) {
+      return this.extractUnstableSubexpressions(node.vExpression, ast);
+    } else if (node instanceof FunctionCall) {
+      return [this.extractIndividualSubExpression(node, ast)];
```

Comment:
Would be interested to see an example of Function(i++) being used as an lvalue or am I missing what is going on here?

Reply:
Oh I tough you were saying `getArray(i++)`